tests/unit/test_upload.py

```diff
@@ -543,3 +610,56 @@ def test_returned_file_mapping_correct_for_failed_uploads(
 
         with self.subTest("no failed uploads"):
             self.assertEqual(failed_files, expected_failed_files)
+
+    @patch("s3_upload.utils.upload.as_completed")
+    @patch("s3_upload.utils.upload._submit_to_pool")
+    @patch("s3_upload.utils.upload.ProcessPoolExecutor")
+    def test_exception_correctly_handled_from_parent_process(
+        self, mock_pool, mock_submit, mock_completed
+    ):
+        """
+        Test that if an error occurs in one of the main ProcessPoolExecutors
+        and that we catch this to not actually exit, the error will be logged
+        and failed set of upload files returned for alerting via Slack and
+        retrying on next call
+        """
+        # set a submitted future to mock a single core upload, we will
+        # raise an exception for when accessing its result object
+        submitted_futures = [Mock()]
+        submitted_futures[0].result.side_effect = RuntimeError(
+            "unhandled test exception"
+        )
+        mock_completed.return_value = submitted_futures
+
+        mock_submit.return_value = {submitted_futures[0]: self.local_files}
+
+        with self.subTest("Error correctly caught and logged"):
+            expected_error = (
+                "Failed uploading 3 files for a single ProcessPool due to an"
```

Comment:
I think I'm missing this, but where does is this error specified/written in the code?

Reply:
it comes from here: https://github.com/eastgenomics/s3_upload/blob/4b1f0f742d76ebe4590845d8efe5c96fdf2fdd37/s3_upload/utils/upload.py#L426

it will hit this point if any of the ProcessPools errors for any reason, we log the error then return that there was a fail to send an alert in the main monitor function
<!-- Reviewable comment -OCZNj5b0fi8afkDFxj9:b2wubtq -->
<!-- Sent from Reviewable.io -->